### ISSUE TITLE
Unable to build js bindings on M1 Mac

### ISSUE THREAD

**Issue by brochington** [opencog/link-grammar#1272]
I am unable to build the js bindings on my M1 Mac Mini. 

The first issue I came across is that `.so` files don't work on mac, which the `bindings/js/link-parser/build.sh` is expecting to use. I've attempted to use either `liblink-grammer.a` or `liblink-grammar.dylib`. I think this has gotten me a bit further. 

Here are what is logged to the console: 
```
+ emmake make
Making all in link-grammar
  CC       api.lo
  CC       connectors.lo
  CC       dict-common/dialect.lo
  CC       dict-common/dict-common.lo
  CC       dict-common/dict-impl.lo
  CC       dict-common/dict-utils.lo
  CC       dict-common/file-utils.lo
  CC       dict-common/idiom.lo
  CC       dict-common/print-dict.lo
  CC       dict-common/regex-morph.lo
  CC       dict-file/read-dialect.lo
  CC       dict-file/dictionary.lo
  CC       dict-file/read-dict.lo
  CC       dict-file/read-regex.lo
  CC       dict-file/word-file.lo
  CC       dict-sql/read-sql.lo
  CC       disjunct-utils.lo
  CC       error.lo
  CC       linkage/analyze-linkage.lo
  CC       linkage/freeli.lo
  CC       linkage/linkage.lo
  CC       linkage/lisjuncts.lo
  CC       linkage/sane.lo
  CC       linkage/score.lo
  CC       memory-pool.lo
  CC       parse/count.lo
  CC       parse/extract-links.lo
  CC       parse/fast-match.lo
  CC       parse/parse.lo
  CC       parse/preparation.lo
  CC       parse/prune.lo
  CC       post-process/constituents.lo
  CC       post-process/post-process.lo
  CC       post-process/pp_knowledge.lo
  CC       post-process/pp_lexer.lo
  CC       post-process/pp_linkset.lo
  CC       prepare/build-disjuncts.lo
  CC       prepare/exprune.lo
  CC       print/print.lo
  CC       print/print-util.lo
  CC       resources.lo
  CC       string-set.lo
  CC       string-id.lo
  CC       tokenize/anysplit.lo
  CC       tokenize/tokenize.lo
  CC       tokenize/wg-display.lo
  CC       tokenize/wordgraph.lo
  CC       tracon-set.lo
  CC       utilities.lo
  CXXLD    liblink-grammar.la
emcc:WARNING: parse/.libs/histogram.o is not a valid input file
emcc:WARNING: print/.libs/wcwidth.o is not a valid input file
emcc:WARNING: tokenize/.libs/spellcheck-aspell.o is not a valid input file
emcc:WARNING: tokenize/.libs/spellcheck-hun.o is not a valid input file
emcc:WARNING: When Emscripten compiles to a typical native suffix for shared libraries (.so, .dylib, .dll) then it emits an object file. You should then compile that to an emscripten SIDE_MODULE (using that flag) with suffix .wasm (for wasm) or .js (for asm.js). (You may also want to adapt your build system to emit the more standard suffix for a an object file, '.bc' or '.o', which would avoid this warning.)
error: cannot parse the debug map for '.libs/liblink-grammar.5.dylib': The file was not recognized as a valid object file
Making all in bindings
Making all in ocaml
make[2]: Nothing to be done for `all'.
make[2]: Nothing to be done for `all-am'.
Making all in link-parser
  CC       link_parser-link-parser.o
  CC       link_parser-command-line.o
  CC       link_parser-lg_readline.o
  CC       link_parser-parser-utilities.o
  CCLD     link-parser
emcc:WARNING: ignoring dynamic library liblink-grammar.5.dylib because not compiling to JS or HTML, remember to link it when compiling to JS or HTML at the end
Making all in tests
make[1]: Nothing to be done for `all'.
Making all in morphology
make[1]: Nothing to be done for `all'.
Making all in data
Making all in ady
make[2]: Nothing to be done for `all'.
Making all in amy
make[2]: Nothing to be done for `all'.
Making all in any
make[2]: Nothing to be done for `all'.
Making all in gen
make[2]: Nothing to be done for `all'.
Making all in ar
Making all in words
make[3]: Nothing to be done for `all'.
make[3]: Nothing to be done for `all-am'.
Making all in de
make[2]: Nothing to be done for `all'.
Making all in en
Making all in words
make[3]: Nothing to be done for `all'.
make[3]: Nothing to be done for `all-am'.
Making all in fa
Making all in words
make[3]: Nothing to be done for `all'.
make[3]: Nothing to be done for `all-am'.
Making all in he
make[2]: Nothing to be done for `all'.
Making all in id
make[2]: Nothing to be done for `all'.
Making all in kz
make[2]: Nothing to be done for `all'.
Making all in lt
make[2]: Nothing to be done for `all'.
Making all in ru
Making all in words
make[3]: Nothing to be done for `all'.
make[3]: Nothing to be done for `all-am'.
Making all in tr
make[2]: Nothing to be done for `all'.
Making all in vn
make[2]: Nothing to be done for `all'.
Making all in demo-sql
make[2]: Nothing to be done for `all'.
make[2]: Nothing to be done for `all-am'.
Making all in man
make[1]: Nothing to be done for `all'.
make[1]: Nothing to be done for `all-am'.
+ popd
~/Downloads/link-grammar-5.10.2/bindings/js ~/Downloads/link-grammar-5.10.2/bindings/js
+ ./link-parser/build.sh
++ dirname ./link-parser/build.sh
+ pushd ./link-parser
~/Downloads/link-grammar-5.10.2/bindings/js/link-parser ~/Downloads/link-grammar-5.10.2/bindings/js
+ '[' '!' /Users/brochstilley/Downloads/link-grammar-5.10.2/bindings/js/emsdk ']'
+ rm -rf dist
+ mkdir dist
+ cp ../../../link-parser/.libs/link-parser link-parser.bc
+ emcc -O3 link-parser.bc ../../../link-grammar/.libs/liblink-grammar.dylib --pre-js pre.js -s WASM=1 -s ALLOW_MEMORY_GROWTH=1 -o dist/link-parser.js
cache:INFO: generating system library: libc-extras.a... (this will be cached in "/Users/brochstilley/Downloads/link-grammar-5.10.2/bindings/js/emsdk/fastcomp/emscripten/cache/wasm-obj/libc-extras.a" for subsequent builds)
cache:INFO:  - ok
error: undefined symbol: mk_wcwidth
warning: To disable errors for undefined symbols use `-s ERROR_ON_UNDEFINED_SYMBOLS=0`
Error: Aborting compilation due to previous errors
shared:ERROR: '/Users/brochstilley/Downloads/link-grammar-5.10.2/bindings/js/emsdk/node/14.15.5_64bit/bin/node /Users/brochstilley/Downloads/link-grammar-5.10.2/bindings/js/emsdk/fastcomp/emscripten/src/compiler.js /var/folders/lc/sjydg8f515j8jj72l376jx1h0000gn/T/tmp8cZ3K_.txt /Users/brochstilley/Downloads/link-grammar-5.10.2/bindings/js/emsdk/fastcomp/emscripten/src/library_pthread_stub.js' failed (1)
```

**Comment by linas** [comment]
Some wild guesses:
> emcc:WARNING: parse/.libs/histogram.o is not a valid input file

is probably due to it no being reccompiled. Try this patch:
```
--- a/bindings/js/build_packages.sh
+++ b/bindings/js/build_packages.sh
@@ -8,6 +8,7 @@ pushd "$(dirname "$0")"
 
 # Compile LLVM bitcode
 pushd ../..
+make clean
 emconfigure ./configure --disable-editline --disable-sat-solver --disable-java-bindings --disable-python-bindings
 emmake make
 popd
```

**Comment by linas** [comment]
And this error:
> error: undefined symbol: mk_wcwidth
is due to this warning:
> emcc:WARNING: print/.libs/wcwidth.o is not a valid input file
because `wcwdith.c` was not recompiled, presumably due to the missing `make clean`

**Comment by ampli** [comment]
> emcc:WARNING: parse/.libs/histogram.o is not a valid input file
emcc:WARNING: print/.libs/wcwidth.o is not a valid input file
emcc:WARNING: tokenize/.libs/spellcheck-aspell.o is not a valid input file
emcc:WARNING: tokenize/.libs/spellcheck-hun.o is not a valid input file

For all the files the warning is correct because they don't contain any symbols (because their source code is #ifdef`ed out).
The reason for the **repeated** problem with `wcwidth.o` is not clear to me.

**Comment by linas** [comment]
@ampli  The code in `wcwidth.c` is called by several utf8 functions in `print-util.c` which are called by assorted print routines. It has to be compiled, and it should contain symbols.

**Comment by brochington** [comment]
I've been able to get this to compile. 

steps:
1. start from fresh tarball
2. `./configure --disable-java-bindings`, then `make` and `makecheck`.
3. add the `make clean` to `build_packages.sh` as you said.
4. in `js/link-parser/build.sh` change `liblink-grammar.so` to `liblink-grammar.a` .
5. run `install_emsdk.sh` and `./build_packages.sh` as specified.

**Comment by linas** [comment]
Care to create a diff, a pull req, with those changes in it?

**Comment by linas** [comment]
Never mind, I figued it out. Fixed in #1305 ... closing.
